types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,518 @@
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(req);
+		return;
+	}
+
+	req->device = device;
+	req->resource = wl_resource;
+
+	wl_resource_set_implementation(wl_resource, &lease_request_impl,
+		req, drm_lease_request_v1_handle_resource_destroy);
+
+	wl_list_insert(&device->requests, wl_resource_get_link(wl_resource));
+}
+
+static struct wp_drm_lease_device_v1_interface lease_device_impl = {
+	.stop = drm_lease_device_v1_handle_stop,
+	.create_lease_request = drm_lease_device_v1_handle_create_lease_request,
+};
+
+static void drm_connector_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+}
+
+static void drm_connector_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_connector_v1_interface lease_connector_impl = {
+	.destroy = drm_connector_v1_handle_destroy,
+};
+
+static void drm_lease_connector_v1_send_to_client(
+		struct wlr_drm_lease_connector_v1 *connector,
+		struct wl_client *wl_client, struct wl_resource *device) {
+	if (connector->active_lease) {
+		return;
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(wl_client,
+			&wp_drm_lease_connector_v1_interface, 1, 0);
+	wl_resource_set_implementation(wl_resource, &lease_connector_impl,
+			connector, drm_connector_v1_handle_resource_destroy);
+	wp_drm_lease_device_v1_send_connector(device, wl_resource);
+
+	struct wlr_output *output = connector->output;
+	wp_drm_lease_connector_v1_send_name(wl_resource, output->name);
+
+	char description[128];
+	snprintf(description, sizeof(description), "%s %s %s (%s)",
+		output->make, output->model, output->serial, output->name);
+	wp_drm_lease_connector_v1_send_description(wl_resource, description);
+
+	wp_drm_lease_connector_v1_send_connector_id(
+			wl_resource, connector->drm_connector->id);
+
+	wl_list_insert(&connector->resources, wl_resource_get_link(wl_resource));
+}
+
+static void lease_device_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *lease_device = data;
+
+	struct wl_resource *wl_resource  = wl_resource_create(wl_client,
+		&wp_drm_lease_device_v1_interface, version, id);
+
+	if (!wl_resource) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(wl_resource, &lease_device_impl,
+		lease_device, drm_lease_device_v1_handle_resource_destroy);
+
+	char *path = drmGetDeviceNameFromFd2(lease_device->backend->fd);
+	int fd = open(path, O_RDWR);
+	if (fd < 0) {
+		wlr_log_errno(WLR_ERROR, "Unable to clone DRM fd for leasing.\n");
+		free(path);
+	} else {
+		if (drmIsMaster(fd)) {
+			drmDropMaster(fd);
+		}
+		assert(!drmIsMaster(fd) && "Won't send master fd to client");
+		wp_drm_lease_device_v1_send_drm_fd(wl_resource, fd);
+		free(path);
+		close(fd);
+	}
```

Comment:
It would be helpful if this could become part of the interface in `include/wlr/backend/drm.h`.

Something like a function `int lease_fd(struct wlr_backend *backend)` which has internally already done all necessary calls into libdrm handing back a cloned non-master fd (and on error would return negative value).

Having it there also feels a bit better in regards of separation of layers as we then wouldn't do calls into libdrm in the frontend.

Reply:
Fixed in https://github.com/swaywm/wlroots/pull/2929/commits/ab35895051c8ec756377c91ba725c549b630a413